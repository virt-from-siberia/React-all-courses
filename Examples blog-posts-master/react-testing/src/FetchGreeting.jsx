import { useState } from 'react'
import axios from 'axios'
import { useGreetingContext } from './GreetingProvider'

const FetchGreeting = ({ url }) => {
  const { state, actions } = useGreetingContext()
  const [btnClicked, setBtnClicked] = useState(false)

  const fetchGreeting = (url) =>
    axios
      .get(url)
      .then((res) => {
        const { data } = res
        const { greeting } = data
        actions.setSuccess(greeting)
        setBtnClicked(true)
      })
      .catch((e) => {
        actions.setError(e)
      })

  const btnText = btnClicked ? 'Готово' : 'Получить приветствие'

  return (
    <div>
      <button onClick={() => fetchGreeting(url)} disabled={btnClicked}>
        {btnText}
      </button>
      {state.greeting && <h1 data-cy='heading'>{state.greeting}</h1>}
      {state.error && <p role='alert'>Не удалось получить приветствие</p>}
    </div>
  )
}

export default FetchGreeting
