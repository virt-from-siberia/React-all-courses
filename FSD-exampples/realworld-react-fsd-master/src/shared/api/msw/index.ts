export { server } from './server';
export { initTestDatabase } from './serverDatabase';
export * from './lib';
