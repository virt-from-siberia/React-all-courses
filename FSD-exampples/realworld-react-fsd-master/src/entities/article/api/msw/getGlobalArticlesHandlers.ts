import { rest } from 'msw';
import {
  server,
  initTestDatabase,
  parseTokenFromRequest,
  mapMswArticleDto,
} from '~shared/api/msw';
import { realworldApi } from '~shared/api/realworld';

const databaseApi = initTestDatabase();

const getGlobalArticlesHandlers = [
  rest.get(`${realworldApi.baseUrl}/articles`, (req, res, ctx) => {
    const author = req.url.searchParams.get('author');
    const favorited = req.url.searchParams.get('favorited');
    const tag = req.url.searchParams.get('tag');
    const offset = Number(req.url.searchParams.get('offset'));
    const limit = Number(req.url.searchParams.get('limit'));

    /**
     * temporary
     * @see https://github.com/mswjs/data/issues/249
     * we have to use databaseApi.article.findMany({ where: {...} })
     */
    const articles = databaseApi.article
      .getAll()
      .filter((article) => (author ? article.authorId === author : true))
      .filter((article) =>
        favorited ? article.favoritedBy.includes(favorited) : true,
      )
      .filter((article) => (tag ? article.tagList.includes(tag) : true));

    const token = parseTokenFromRequest(req);
    const maybeUser = databaseApi.user.findFirst({
      where: { token: { equals: token } },
    });

    const articlesDto = articles
      .slice(offset, limit + offset)
      .map((article) => {
        const maybeProfile = databaseApi.profile.findFirst({
          where: { username: { equals: article.authorId } },
        });

        return mapMswArticleDto(article, maybeUser, maybeProfile);
      });

    return res(
      ctx.status(200),
      ctx.json({
        articles: articlesDto,
        articlesCount: articles.length,
      }),
    );
  }),
];

export const setupGetGlobalArticlesHandlers = () => {
  server.use(...getGlobalArticlesHandlers);
};
