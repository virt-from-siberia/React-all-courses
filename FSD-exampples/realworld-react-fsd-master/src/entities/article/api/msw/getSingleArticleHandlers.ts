import { rest } from 'msw';
import {
  server,
  initTestDatabase,
  parseTokenFromRequest,
  mapMswArticleDto,
} from '~shared/api/msw';
import { realworldApi } from '~shared/api/realworld';

const databaseApi = initTestDatabase();

const getSingleArticleHandlers = [
  rest.get(`${realworldApi.baseUrl}/articles/:slug`, (req, res, ctx) => {
    const { slug } = req.params;

    const maybeArticle = databaseApi.article.findFirst({
      where: { slug: { equals: String(slug) } },
    });

    if (!maybeArticle)
      return res(
        ctx.status(404),
        ctx.json({ errors: { article: ['not found'] } }),
      );

    const token = parseTokenFromRequest(req);
    const maybeUser = databaseApi.user.findFirst({
      where: { token: { equals: token } },
    });

    const maybeProfile = databaseApi.profile.findFirst({
      where: { username: { equals: maybeArticle.authorId } },
    });

    const articleDto = mapMswArticleDto(maybeArticle, maybeUser, maybeProfile);

    return res(ctx.status(200), ctx.json({ article: articleDto }));
  }),
];

export const setupGetSingleArticleHandlers = () => {
  server.use(...getSingleArticleHandlers);
};
