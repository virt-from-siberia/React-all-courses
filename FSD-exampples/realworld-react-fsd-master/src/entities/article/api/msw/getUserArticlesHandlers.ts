import { rest } from 'msw';
import {
  server,
  initTestDatabase,
  parseTokenFromRequest,
  mapMswArticleDto,
} from '~shared/api/msw';
import { realworldApi } from '~shared/api/realworld';

const databaseApi = initTestDatabase();

const getUserArticlesHandlers = [
  rest.get(`${realworldApi.baseUrl}/articles/feed`, (req, res, ctx) => {
    const offset = Number(req.url.searchParams.get('offset'));
    const limit = Number(req.url.searchParams.get('limit'));

    const token = parseTokenFromRequest(req);
    const maybeUser = databaseApi.user.findFirst({
      where: { token: { equals: token } },
    });

    if (!maybeUser)
      return res(
        ctx.status(401),
        ctx.json({
          status: 'error',
          message: 'missing authorization credentials',
        }),
      );

    const articles = databaseApi.article.findMany({
      where: { authorId: { equals: maybeUser.username } },
      take: limit,
      skip: offset,
    });
    const articlesCount = databaseApi.article.count({
      where: { authorId: { equals: maybeUser.username } },
    });

    const articlesDto = articles.map((article) => {
      const maybeProfile = databaseApi.profile.findFirst({
        where: { username: { equals: article.authorId } },
      });

      return mapMswArticleDto(article, maybeUser, maybeProfile);
    });

    return res(
      ctx.status(200),
      ctx.json({
        articles: articlesDto,
        articlesCount,
      }),
    );
  }),
];

export const setupGetUserArticlesHandlers = () => {
  server.use(...getUserArticlesHandlers);
};
