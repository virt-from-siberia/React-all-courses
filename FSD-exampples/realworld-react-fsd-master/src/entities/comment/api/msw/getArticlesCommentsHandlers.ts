import { rest } from 'msw';
import {
  server,
  initTestDatabase,
  parseTokenFromRequest,
  mapMswProfileDto,
} from '~shared/api/msw';
import { realworldApi } from '~shared/api/realworld';

const databaseApi = initTestDatabase();

const getArticlesCommentsHandlers = [
  rest.get(
    `${realworldApi.baseUrl}/articles/:slug/comments`,
    (req, res, ctx) => {
      const { slug } = req.params;

      const comments = databaseApi.comment.findMany({
        where: { articleId: { equals: String(slug) } },
      });

      const token = parseTokenFromRequest(req);
      const maybeUser = databaseApi.user.findFirst({
        where: { token: { equals: token } },
      });

      const commentsDto = comments.map((curComment) => {
        const { articleId, authorId, ...comment } = curComment;

        const maybeProfile = databaseApi.profile.findFirst({
          where: { username: { equals: authorId } },
        });

        const author = mapMswProfileDto(maybeUser, maybeProfile);

        return {
          ...comment,
          author,
        };
      });

      return res(ctx.status(200), ctx.json({ comments: commentsDto }));
    },
  ),
];

export const setupGetArticlesCommentsHandlers = () => {
  server.use(...getArticlesCommentsHandlers);
};
