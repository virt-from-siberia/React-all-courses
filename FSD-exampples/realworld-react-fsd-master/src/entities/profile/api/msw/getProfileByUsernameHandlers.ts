import { rest } from 'msw';
import {
  server,
  initTestDatabase,
  parseTokenFromRequest,
  mapMswProfileDto,
} from '~shared/api/msw';
import { realworldApi } from '~shared/api/realworld';

const databaseApi = initTestDatabase();

const getProfileByUsernameHandlers = [
  rest.get(`${realworldApi.baseUrl}/profiles/:username`, (req, res, ctx) => {
    const { username } = req.params;

    const token = parseTokenFromRequest(req);
    const maybeUser = databaseApi.user.findFirst({
      where: { token: { equals: token } },
    });

    const maybeProfile = databaseApi.profile.findFirst({
      where: { username: { equals: String(username) } },
    });

    if (!maybeProfile)
      return res(
        ctx.status(404),
        ctx.json({ errors: { profile: ['not found'] } }),
      );

    const profileDto = mapMswProfileDto(maybeUser, maybeProfile);

    return res(ctx.status(200), ctx.json({ profile: profileDto }));
  }),
];

export const setupGetProfileByUsernameHandlers = () => {
  server.use(...getProfileByUsernameHandlers);
};
