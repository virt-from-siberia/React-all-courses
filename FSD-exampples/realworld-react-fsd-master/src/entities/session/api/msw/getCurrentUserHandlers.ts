import { rest } from 'msw';
import {
  server,
  initTestDatabase,
  parseTokenFromRequest,
} from '~shared/api/msw';
import { realworldApi } from '~shared/api/realworld';

const databaseApi = initTestDatabase();

const getCurrentUserHandlers = [
  rest.get(`${realworldApi.baseUrl}/user`, (req, res, ctx) => {
    const token = parseTokenFromRequest(req);

    if (!token)
      return res(
        ctx.status(401),
        ctx.json({
          status: 'error',
          message: 'missing authorization credentials',
        }),
      );

    const maybeUser = databaseApi.user.findFirst({
      where: { token: { equals: token } },
    });

    if (maybeUser) return res(ctx.status(200), ctx.json({ user: maybeUser }));

    return res(
      ctx.status(401),
      ctx.json({
        status: 'error',
        message: 'missing authorization credentials',
      }),
    );
  }),
];

export const setupGetCurrentUserHandlers = () => {
  server.use(...getCurrentUserHandlers);
};
