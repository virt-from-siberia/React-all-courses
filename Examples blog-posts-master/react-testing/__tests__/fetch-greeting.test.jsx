import { rest } from 'msw'
import { setupServer } from 'msw/node'
import userEvent from '@testing-library/user-event'
import { render, fireEvent, waitFor, screen } from 'test-utils'
import '@testing-library/jest-dom'
import FetchGreeting from '../src/FetchGreeting'

const server = setupServer(
  rest.get('/greeting', (req, res, ctx) =>
    res(ctx.json({ greeting: 'Привет!' }))
  )
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('получение приветствия', () => {
  test('-> успешное получение и отображение приветствия', async function () {
    const { container, getByDataCy } = render(<FetchGreeting url='/greeting' />)
    expect(container).toMatchSnapshot()

    fireEvent.click(screen.getByText('Получить приветствие'))

    await waitFor(() => screen.getByRole('heading'))
    expect(container).toMatchSnapshot()

    expect(getByDataCy('heading')).toHaveTextContent('Привет!')
    expect(screen.getByRole('button')).toHaveTextContent('Готово')
    expect(screen.getByRole('button')).toBeDisabled()
  })

  it('-> обработка ошибки сервера', async () => {
    server.use(rest.get('/greeting', (req, res, ctx) => res(ctx.status(500))))

    const { container } = render(<FetchGreeting url='greeting' />)
    expect(container).toMatchSnapshot()

    const user = userEvent.setup()
    await user.click(screen.getByText('Получить приветствие'))

    await waitFor(() => screen.getByRole('alert'))
    expect(container).toMatchSnapshot()

    expect(screen.getByRole('alert')).toHaveTextContent(
      'Не удалось получить приветствие'
    )
    expect(screen.getByRole('button')).not.toBeDisabled()
  })
})
