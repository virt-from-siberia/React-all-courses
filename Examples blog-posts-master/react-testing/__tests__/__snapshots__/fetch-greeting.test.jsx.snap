// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`получение приветствия -> обработка ошибки сервера 1`] = `
<div>
  <div>
    <button>
      Получить приветствие
    </button>
  </div>
</div>
`;

exports[`получение приветствия -> обработка ошибки сервера 2`] = `
<div>
  <div>
    <button>
      Получить приветствие
    </button>
    <p
      role="alert"
    >
      Не удалось получить приветствие
    </p>
  </div>
</div>
`;

exports[`получение приветствия -> успешное получение и отображение приветствия 1`] = `
<div>
  <div>
    <button>
      Получить приветствие
    </button>
  </div>
</div>
`;

exports[`получение приветствия -> успешное получение и отображение приветствия 2`] = `
<div>
  <div>
    <button
      disabled=""
    >
      Готово
    </button>
    <h1
      data-cy="heading"
    >
      Привет!
    </h1>
  </div>
</div>
`;
